include/aspect/stokes_matrix_free.h

```diff
@@ -274,11 +274,72 @@ namespace aspect
   }
 
   /**
-   * Main class of the Matrix-free method. Here are all the functions for setup, assembly and solving
-   * the Stokes system.
-   */
+    * Base class for the matrix free GMG solver for the Stokes system. The
+    * actual implementation is found inside StokesMatrixFreeHandlerImpl below.
+    */
   template<int dim>
   class StokesMatrixFreeHandler
+  {
+    public:
+      /**
+       * Destructor.
+       */
+      virtual ~StokesMatrixFreeHandler();
+
+      /**
+       * Solves the Stokes linear system matrix-free. This is called
+       * by Simulator<dim>::solve_stokes().
+       */
+      virtual std::pair<double,double> solve()=0;
+
+      /**
+       * Allocates and sets up the members of the StokesMatrixFreeHandler. This
+       * is called by Simulator<dim>::setup_dofs()
+       */
+      virtual void setup_dofs()=0;
+
+      /**
+       * Evalute the MaterialModel to query for the viscosity on the active cells,
+       * project this viscosity to the multigrid hierarchy, and cache the information
+       * for later usage. Also sets pressure scaling and information regarding the
+       * compressiblity of the flow.
+       */
+      virtual void evaluate_material_model()=0;
+
+      /**
+       * Get the workload imbalance of the distribution
+       * of the level hierarchy.
+       */
+      virtual double get_workload_imbalance()=0;
+
+      /**
+       * Add correction to system RHS for non-zero boundary condition.
+       */
+      virtual void correct_stokes_rhs()=0;
+
+      /**
+       * Declare parameters. (No actual parameters at the moment).
+       */
+      static
+      void declare_parameters (ParameterHandler &prm);
+
+      /**
+       * Parse parameters. (No actual parameters at the moment).
+       */
+      virtual void parse_parameters (ParameterHandler &prm);
+  };
+
+  /**
+   * Main class of the Matrix-free method. Here are all the functions for
+   * setup, assembly and solving the Stokes system.
+   *
+   * We need to derive from StokesMatrixFreeHandler to be able to introduce a
+   * second template argument for the degree of the Stokes finite
+   * element. This way, the main simulator does not need to know about the
+   * degree and we can pick the desired class to use at runtime.
+   */
+  template<int dim, int velocity_degree>
+  class StokesMatrixFreeHandlerImpl: public StokesMatrixFreeHandler<dim>
   {
```

Comment:
To me `StokesMatrixFreeHandlerImpl` looks like `StokesMatrixFreeHandlerImp1`. Is ` StokesMatrixFreeHandlerImplemntation` too long? Maybe we could remove the word "Handler"?  Or maybe we call the base class `StokesMatrixFreeHandlerBase`?

Reply:
> Is StokesMatrixFreeHandlerImplemntation` too long?

no, I think this would be fine to use.